[memory] Add torch memory snapshot profiling capabilities

minosfuture opened the pull request:
## Purpose


  This PR adds memory snapshot profiling capabilities to vLLM, enabling detailed analysis of GPU memory allocation during model initialization and inference. Memory snapshots can be
  visualized using https://pytorch.org/memory_viz.

This should also be useful for debugging GPU OOM.

  Key features:

  - --profiler-config interface command-line args
  - API endpoints: /start_mem_profile and /stop_mem_profile for on-demand profiling during inference
  - Automatic init profiling: Captures memory snapshots for key initialization stages:
    - load_model - Model weights loading
    - kv_cache - KV cache allocation
    - cuda_graph - CUDA graph capture
  - Exception handling: Automatically dumps snapshot when exceptions occur (configurable)

  Configuration options:

  | Option                            | Default | Description                                                 |
  |-----------------------------------|---------|-------------------------------------------------------------|
  | memory_profiler_dir               | ""      | Output directory for snapshots (enables profiling when set) |
  | memory_profiler_max_entries       | 100000  | Max allocation entries to record                            |
  | memory_profiler_profile_init      | True    | Auto-profile during initialization                          |
  | memory_profiler_dump_on_exception | True    | Dump snapshot on exception                                  |

  Example usage:
```
 vllm serve deepseek-ai/DeepSeek-V2-Lite --profiler-config.memory_profiler_dir=/tmp/mem_snapshots
```
  Then trigger on-demand profiling:
``` 
curl -X POST http://localhost:8000/start_mem_profile
# ... run inference ...
curl -X POST http://localhost:8000/stop_mem_profile
```

## Test Plan

  - Unit tests for MemorySnapshotProfiler class 
  - Unit tests for API router endpoints
  - All tests pass: pytest tests/v1/worker/test_memory_profiler.py tests/entrypoints/test_profile_api.py -v
  - Manual testing with actual model loading

`vllm serve deepseek-ai/DeepSeek-V2-Lite --profiler-config.memory_profiler_dir=/tmp/mem_snapshots`

## Test Result
```
(EngineCore_DP0 pid=2781349) INFO 12-19 22:25:34 [mem_utils.py:371] Memory snapshot saved to /tmp/mem_snapshots/load_model_rank0_20251219_222534.pickle. Visualize at https://pytorch.org/memory_viz
(EngineCore_DP0 pid=2781349) INFO 12-19 22:25:39 [mem_utils.py:371] Memory snapshot saved to /tmp/mem_snapshots/kv_cache_rank0_20251219_222539.pickle. Visualize at https://pytorch.org/memory_viz
(EngineCore_DP0 pid=2781349) INFO 12-19 22:25:42 [mem_utils.py:371] Memory snapshot saved to /tmp/mem_snapshots/cuda_graph_rank0_20251219_222542.pickle. Visualize at https://pytorch.org/memory_viz
```

Example:

<img width="2558" height="1203" alt="image" src="https://github.com/user-attachments/assets/454e98b8-e591-4dce-ab28-c9be6752aa85" />


---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

minosfuture committed:
[memory] Add torch memory snapshot dump

Signed-off-by: Ming Yang <minos.future@gmail.com>

mgoin commented:
Does the memory trace allow tracking where in the code each allocation was made? This would be useful to see what are weights, kernel workspaces, activations, etc during multiple phases of vllm's init and runtime

MatthewBonanni commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -172,6 +176,17 @@ def _maybe_get_memory_pool_context(self, tag: str) -> AbstractContextManager:
+    def _maybe_get_memory_snapshot_context(self) -> AbstractContextManager:
+        # Optional memory snapshot profiling for debugging OOM during loading
+        memory_snapshot_ctx: AbstractContextManager = nullcontext()
+        if envs.VLLM_MEMORY_SNAPSHOT_DIR:
```
Could we use the `--profiler-config` interface instead of an environment variable?

MatthewBonanni reviewed (commented):
This is cool, thanks! Would it be reasonable to enable it for the full startup sequence? Or even for inference as well? It would be really useful to have broader insight into the memory footprint.

Relatedly, I'm working on improving the startup memory profiling in #30515

minosfuture commented:
> This is cool, thanks! Would it be reasonable to enable it for the full startup sequence? Or even for inference as well? It would be really useful to have broader insight into the memory footprint.
> 
> 
> 
> Relatedly, I'm working on improving the startup memory profiling in #30515

Sounds good. Lemme extend the coverage:
1. Expose /start_mem_profile, /stop_mem_profile endpoints. UX should be similar to torch profile.
2. When enabled, dump snapshot upon exception.
3. Dump snapshots for model loading and whole init stage.

Another extension I'll try is to post process the pickle file and aggregate the profile into a flame graph.

MatthewBonanni commented:
Sounds fantastic, that's exactly what I was starting to hack together locally once I saw how useful this was :) Thanks!

minosfuture committed:
Merge remote-tracking branch 'origin/main' into memory_profile

minosfuture committed:
[memory] Add memory snapshot profiler with API endpoints

- Add MemorySnapshotProfiler class with start/stop API and context manager support
- Add /start_mem_profile and /stop_mem_profile API endpoints
- Add memory profiler settings to ProfilerConfig:
  - memory_profiler_dir: Directory for snapshots
  - memory_profiler_max_entries: Max entries to record
  - memory_profiler_profile_init: Profile during model loading
  - memory_profiler_dump_on_exception: Dump on errors
- Remove VLLM_MEMORY_SNAPSHOT_DIR and VLLM_MEMORY_SNAPSHOT_MAX_ENTRIES env vars
- Wire up mem_profile method through engine chain

Snapshots can be visualized at https://pytorch.org/memory_viz

Usage:
  # Enable profiler endpoints
  --profiler-config.memory_profiler_dir=/path/to/output

  # Auto-profile during init
  --profiler-config.memory_profiler_profile_init=true

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
[memory] Add unit tests for memory snapshot profiler

Add comprehensive tests for:
- MemorySnapshotProfiler class (start/stop, context manager, exception
  handling, filename generation, rank support)
- Profile API endpoints (/start_mem_profile, /stop_mem_profile, router
  attachment logic)

Move tests to tests/v1/worker/test_memory_profiler.py alongside
test_gpu_profiler.py.

Change default of memory_profiler_profile_init to True to enable init
profiling when memory_profiler_dir is set.

Signed-off-by: Ming Yang <minos.future@gmail.com>

minosfuture committed:
[memory] Add memory profiling for kv_cache and cuda_graph stages

Add memory snapshot profiling for additional initialization stages:
- kv_cache: KV cache allocation
- cuda_graph: CUDA graph capture

Signed-off-by: Ming Yang <minos.future@gmail.com>

MatthewBonanni commented:
This looks great, thanks! Could you make it profile the entire startup by default rather than just those stages? There's other stuff during startup that I'd like to analyze.

micah-wil commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -111,6 +115,17 @@ def __init__(
             self.profiler = None
 
+        # Memory snapshot profiler. Enabled when memory_profiler_dir is set.
+        if profiler_config.memory_profiler_enabled:
```
Hi @minosfuture, this causes the following error when running without mem profiling enabled:
```
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751] Traceback (most recent call last):
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]   File "/projects/vllm/vllm/v1/executor/multiproc_executor.py", line 722, in worker_main
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]     worker = WorkerProc(*args, **kwargs)
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]   File "/projects/vllm/vllm/v1/executor/multiproc_executor.py", line 562, in __init__
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]     self.worker.load_model()
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]   File "/projects/vllm/vllm/v1/worker/gpu_worker.py", line 313, in load_model
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]     self._maybe_get_memory_snapshot_context(),
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]   File "/projects/vllm/vllm/utils/mem_utils.py", line 413, in __enter__
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]     self.start()
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]   File "/projects/vllm/vllm/utils/mem_utils.py", line 322, in start
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]     os.makedirs(self.output_dir, exist_ok=True)
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751]   File "<frozen os>", line 225, in makedirs
(Worker_TP3 pid=140477) ERROR 02-06 17:20:21 [multiproc_executor.py:751] FileNotFoundError: [Errno 2] No such file or directory: ''
```

 (repro example: `vllm serve /models/Llama-3.1-405B-Instruct-FP8-KV -tp 8 --dtype float16`)

I think the condition just needs to be changed to something like: 
```
if profiler_config.memory_profiler_profile_init and profiler_config.memory_profiler_dir:
            profiler = MemorySnapshotProfiler( .....
```

Thanks for working on this feature! I am actively using it for a debug
